### ISSUE TITLE
ThingShadowEcho not registering shadow delta function on SDK 2.1.1

### ISSUE THREAD

**Issue by emepetres** [aws/aws-iot-device-sdk-arduino-yun#16]
Hi, I have an Arduino Yún with SDK 2.1.1 installed where I have successfully tested BasicPubSub example with no problems. 

However, ThingShadowEcho example gets stuck after printing "shadow init completed", without any error message. I have configured "aws_iot_config.h" the same way as in BasicPubSub example, so I think everything is in its place.

This is the what the serial monitor shows:
AWS IoT SDK Version(dev) 2.1.1-

[LOG] command: setup completed.
[LOG] command: config completed.
[LOG] command: connect completed.
[LOG] command: shadow init completed.

Any insights? Thanks.

**Comment by liuszeng** [comment]
Hi @emepetres ,

Thank you very much for using AWS IoT Arduino Yun SDK!

The example you are looking at is the ThingShadowEcho sample, which gives you a general idea of how devices are interacting with AWS IoT using the IoT Yun SDK. It listens on the registered shadow delta topic and sync the delta state (the desired state that are different from the reported state) to match the reported state.

To run this sample, you will need to update the desired state (only the desired state) of the same shadow registered from elsewhere, e.g., your [AWS IoT Console](https://console.aws.amazon.com/iot). You should be able to see the shadow reported state updated as well in your AWS IoT Console shadow document once the sketch is running.

The issue you are experiencing here might be caused by serial1 communication latency. We are actively looking into this. Could you try to update the reported state of the shadow in your AWS IoT Console once the sketch is running to see if the shadow JSON document is updated and if there is any output in the serial monitor?

Also, one thing you can try here is to add an extra line: `Serial.flush();` at the end of the logging statement to find out if it is an internal SDK issue or a Serial write buffer issue.

Please let us know how it goes.

Thanks,
Liusu

**Comment by emepetres** [comment]
Hi @liuszeng , thank you for answering.

I tried updating the shadow (only desired state) from the IoT console, but same result. It seems that the code is stuck registering the delta function, and therefore the loop is not running.

However adding `Serial.flush();` before the return inside the print_log function (line 41) makes the example run as expected, updating the shadow. This is the log I get:

```
AWS IoT SDK Version(dev) 2.1.1-

[LOG] command: setup completed.
[LOG] command: config completed.
[LOG] command: connect completed.
[LOG] command: shadow init completed.
[LOG] command: register thing shadow delta function completed.
[LOG] command: getDeltaKeyValue completed.
[LOG] command: update thing shadow completed.
```

So flushing the serial port seems to fix the problem, thanks!

**Comment by liuszeng** [comment]
Hi @emepetres ,

Glad to know that you got the example working. We will incorporate the example changes into the upcoming release. Thank you very much for your effort in making the SDK better!

Thanks,
Liusu

**Comment by liuszeng** [comment]
Fixed in release v2.2.0.
